src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rest/QueryBatchRequestHandler.java

```diff
@@ -0,0 +1,309 @@
+
+    try {
+      // Parse and validate parameters
+      String[] validParamsTmp = {"", "metrics", "starttime", "endtime", "samplingperiod"};
+      Set<String> validParams = new HashSet<>(Arrays.asList(validParamsTmp));
+      for (String param : params.keySet()) {
+        if (!validParams.contains(param)) {
+          throw new InvalidParameterException(String.format("%s is an invalid parameter", param));
+        }
+      }
+
+      List<String> metrics = metricsRestUtil.parseArrayParam(params, "metrics", false);
+      String startTimeParam = params.get("starttime");
+      String endTimeParam = params.get("endtime");
+      String samplingPeriodParam = params.get("samplingperiod");
+
+      for (String metric : metrics) {
+        if (!MetricsModel.ALL_METRICS.containsKey(metric)) {
+          throw new InvalidParameterException(String.format("%s is an invalid metric", metric));
+        }
+      }
+
+      if (startTimeParam == null || startTimeParam.isEmpty()) {
+        throw new InvalidParameterException("starttime parameter must be set");
+      }
+      long startTime;
+      try {
+        startTime = Long.parseUnsignedLong(startTimeParam);
+      } catch (NumberFormatException e) {
+        throw new InvalidParameterException(String.format("%s is an invalid starttime", startTimeParam));
+      }
+
+      if (endTimeParam == null || endTimeParam.isEmpty()) {
+        throw new InvalidParameterException("endtime parameter must be set");
+      }
+      long endTime;
+      try {
+        endTime = Long.parseUnsignedLong(endTimeParam);
+      } catch (NumberFormatException e) {
+        throw new InvalidParameterException(String.format("%s is an invalid endtime", endTimeParam));
+      }
+
+      long samplingPeriod = DEFAULT_SAMPLING_PERIOD_MILLIS;
+      if (samplingPeriodParam != null && !samplingPeriodParam.isEmpty()) {
+        samplingPeriod = Long.parseLong(samplingPeriodParam);
+        if (samplingPeriod < 5 || samplingPeriod % 5 != 0) {
+          throw new InvalidParameterException(String.format("%s is an invalid sampling period", samplingPeriodParam));
+        }
+        if (samplingPeriod >= PluginSettings.instance().getBatchMetricsRetentionPeriodMinutes() * 60) {
+          throw new InvalidParameterException("sampling period must be less than the retention period");
+        }
+        samplingPeriod *= 1000;
+      }
+
+      if (startTime >= endTime) {
+        throw new InvalidParameterException("starttime must be less than the endtime");
+      }
+      startTime -= startTime % samplingPeriod;
+      endTime -= endTime % samplingPeriod;
+      if (startTime == endTime) {
+        throw new InvalidParameterException("starttime and endtime cannot be equal when rounded down to the nearest sampling period");
+      }
+      if (endTime > currentTime) {
+        throw new InvalidParameterException("endtime can be no greater than the system time at the node");
```

Comment:
Minor: System time at node should not matter, right? Since startTime and endTime are provided as utc epoch timestamp values. We can just mention something like expected value is utc seconds since epoch timestamp and provided value is in future.

Reply:
I use the phrase "system time" because we determine current time using System.currentTimeMillis, which in turn uses methods from the operating system. And I believe machines can be setup to indicate whatever timestamp the owner wants -- like, I could setup a computer today with a current time of January 1, 1970. So, I expose this additional detail in the error message to help anyone who might be having problems, while keeping the readme and design doc at the abstraction of utc for the standard use case.